src/main/java/com/zendesk/maxwell/MaxwellAbstractRowsEvent.java

```diff
@@ -275,21 +292,59 @@ public Object getData(String string) {
 		return list;
 	}
 
-	public List<JSONObject> toJSONObjects() {
-		ArrayList<JSONObject> list = new ArrayList<>();
+	static final JsonFactory factory = new JsonFactory();
 
-		for ( RowMap map : jsonMaps() ) {
-			list.add(new MaxwellJSONObject(map));
+	private JsonGenerator createJSONGenerator(ByteArrayOutputStream b) {
+		try {
+			return factory.createGenerator(b);
+		} catch (IOException e) {
 		}
-		return list;
+		return null;
+	}
+
+	private void rowMapToJSON(JsonGenerator g, RowMap map) throws IOException {
+		g.writeStartObject(); // {
+
+		for ( String key: map.keySet() ) {
+			g.writeObjectField(key, map.get(key)); // type: "insert"
+		}
+
+		g.writeObjectFieldStart("data");
+		for ( String key: map.data.keySet() ) {
+			Object data = map.getData(key);
+
+			if ( data != null ) {
+				if ( data instanceof List ) { // sets come back from .asJSON as lists, and jackson can't deal.
+					List<String> stringList = (List<String>) data;
+
+					g.writeArrayFieldStart(key);
+					for ( String s : stringList )  {
+						g.writeString(s);
+					}
+					g.writeEndArray();
+				} else {
+					g.writeObjectField(key, data);
+				}
+			}
+		}
+		g.writeEndObject();
+		g.writeEndObject();
 	}
 
 	public List<String> toJSONStrings() {
 		ArrayList<String> list = new ArrayList<>();
 
 		for ( RowMap map : jsonMaps() ) {
-			list.add(new MaxwellJSONObject(map).toString());
+			ByteArrayOutputStream b = new ByteArrayOutputStream();
+			JsonGenerator g = createJSONGenerator(b);
+
+			try {
+				rowMapToJSON(g, map);
+				g.flush();
+				list.add(b.toString());
+			} catch ( IOException e ) { /* never gonna happen. */ }
```

Comment:
C'mon empty catch clauses is the "goto" of Java
If it ain't never gonna happen, then adding `logger.error` won't hurt anything :)